### ISSUE TITLE
Proxy Container memory usage too high in 0.4.0 db-less mode

### ISSUE THREAD

**Issue by HsinHeng** [Kong/kubernetes-ingress-controller#282]
NOTE: GitHub issues are reserved for **bug reports only**.
For anything else, please join the conversation
in Kong Nation https://discuss.konghq.com/c/kubernetes.

---

### Summary

SUMMARY_GOES_HERE

**Kong Ingress controller version**
0.4.0

**Kong or Kong Enterprise version**
-

### Kubernetes version
1.11

**paste `kubectl version` output**
Client Version: version.Info{Major:"1", Minor:"13", GitVersion:"v1.13.1", GitCommit:"eec55b9ba98609a46fee712359c7b5b365bdd920", GitTreeState:"clean", BuildDate:"2018-12-13T10:39:04Z", GoVersion:"go1.11.2", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"11+", GitVersion:"v1.11.8-eks-7c34c0", GitCommit:"7c34c0d2f2d0f11f397d55a46945193a0e22d8f3", GitTreeState:"clean", BuildDate:"2019-03-01T22:49:39Z", GoVersion:"go1.10.8", Compiler:"gc", Platform:"linux/amd64"}

### Environment

- **Cloud provider or hardware configuration**:
AWS EKS

- **OS** (e.g. from /etc/os-release):
- **Kernel** (e.g. `uname -a`):
- **Install tools**:
- **Others**:

### What happened
I apply DB-less mode to AWS EKS 3 day ago.
Memory usage of proxy container seems use too much memory (not any clients requests.)

### Steps To Reproduce

1. kubectl apply -f https://bit.ly/kong-ingress-dbless.
2. I customize the kong-ingress-dbless.yaml, set kong-admin container port of 0.0.0.0:8001.
2. 3 days later...
3. kubectl top pod ingress-kong-{id} --namespace={namespace} --containers=true

POD                             NAME                 CPU(cores)   MEMORY(bytes)
**ingress-kong-58f79d6645-c7qgh   proxy                0m           1720Mi**
ingress-kong-58f79d6645-c7qgh   ingress-controller   3m           8Mi

This is @ dev phase.  No any client requests to this container.
I found that proxy container used bewteen 1.5G ~ 2.5G memory.
is it used too much?
I check the log almost as bellow..

127.0.0.1 - - [28/Apr/2019:15:12:49 +0000] "POST /config HTTP/1.1" 201 3472 "-" "Go-http-client/1.1"
2019/04/28 15:12:52 [notice] 39#0: *1166836 [lua] cache.lua:186: purge(): [DB cache] purging (local) cache, client: 127.0.0.1, server: kong_admin, request: "POST /config HTTP/1.1", host: "localhost:8444"

A cronjob always use polling to POST /config.

**Comment by julianguinard** [comment]
I had a similar issue on EKS with 2 t3.small instances as worker nodes as well, using a postgres DB as kong's backend.

After a few minutes the node running kong-ingress-controller pod would end up consuming 100% of the CPU, leading to pods going into "NodeLost" status.

Reverting to kong-ingress-controller v0.3.0 seems to fix the problem

**Comment by hbagdi** [comment]
> After a few minutes the node running kong-ingress-controller pod would end up consuming 100% of the CPU

@julianguinard  Could you open a separate Github issue. It seems like your issue is with 100% CPU usage and is different from @HsinHeng's proxy pod taking up high memory.
Please ensure that you include error logs. Also, if you can share the Ingress rules and CRs you're using will be helpful. Thanks!

**Comment by hbagdi** [comment]
@HsinHeng
Thanks for the report. I've a few questions to better understand the problem:
1. Could you share the Ingress specs and CRs that you've defined?
2. How many nginx worker threads are running inside the proxy container? You can find this by execing into the container.
3. Do you have a graph of memory usage by the pod over time?

**Comment by HsinHeng** [comment]
@hbagdi Thanks for your reply.

1. CRs are copied latest version from https://bit.ly/kong-ingress-dbless, 
My yamls as bellows:
```apiVersion: apiextensions.k8s.io/v1beta1
kind: CustomResourceDefinition
metadata:
  name: kongplugins.configuration.konghq.com
spec:
  group: configuration.konghq.com
  version: v1
  scope: Namespaced
  names:
    kind: KongPlugin
    plural: kongplugins
    shortNames:
    - kp
  additionalPrinterColumns:
  - name: Plugin-Type
    type: string
    description: Name of the plugin
    JSONPath: .plugin
  - name: Age
    type: date
    description: Age
    JSONPath: .metadata.creationTimestamp
  - name: Disabled
    type: boolean
    description: Indicates if the plugin is disabled
    JSONPath: .disabled
    priority: 1
  - name: Config
    type: string
    description: Configuration of the plugin
    JSONPath: .config
    priority: 1
  validation:
    openAPIV3Schema:
      required:
      - plugin
      properties:
        plugin:
          type: string
        disabled:
          type: boolean
        config:
          type: object
        run_on:
          type: string
          enum:
          - first
          - second
          - all
        protocols:
          type: array
          items:
            type: string
            enum:
            - http
            - https
            - tcp
            - tls


---

apiVersion: apiextensions.k8s.io/v1beta1
kind: CustomResourceDefinition
metadata:
  name: kongconsumers.configuration.konghq.com
spec:
  group: configuration.konghq.com
  version: v1
  scope: Namespaced
  names:
    kind: KongConsumer
    plural: kongconsumers
    shortNames:
    - kc
  additionalPrinterColumns:
  - name: Username
    type: string
    description: Username of a Kong Consumer
    JSONPath: .username
  - name: Age
    type: date
    description: Age
    JSONPath: .metadata.creationTimestamp
  validation:
    openAPIV3Schema:
      properties:
        username:
          type: string
        custom_id:
          type: string

---

apiVersion: apiextensions.k8s.io/v1beta1
kind: CustomResourceDefinition
metadata:
  name: kongcredentials.configuration.konghq.com
spec:
  group: configuration.konghq.com
  version: v1
  scope: Namespaced
  names:
    kind: KongCredential
    plural: kongcredentials
  additionalPrinterColumns:
  - name: Credential-type
    type: string
    description: Type of credential
    JSONPath: .type
  - name: Age
    type: date
    description: Age
    JSONPath: .metadata.creationTimestamp
  - name: Consumer-Ref
    type: string
    description: Owner of the credential
    JSONPath: .consumerRef
  validation:
    openAPIV3Schema:
      required:
      - consumerRef
      - type
      properties:
        consumerRef:
          type: string
        type:
          type: string

---

apiVersion: apiextensions.k8s.io/v1beta1
kind: CustomResourceDefinition
metadata:
  name: kongingresses.configuration.konghq.com
spec:
  group: configuration.konghq.com
  version: v1
  scope: Namespaced
  names:
    kind: KongIngress
    plural: kongingresses
    shortNames:
    - ki
  validation:
    openAPIV3Schema:
      properties:
        upstream:
          type: object
        route:
          properties:
            methods:
              type: array
              items:
                type: string
            regex_priority:
              type: integer
            strip_path:
              type: boolean
            preserve_host:
              type: boolean
            protocols:
              type: array
              items:
                type: string
                enum:
                - http
                - https
        proxy:
          type: object
          properties:
            protocol:
              type: string
              enum:
              - http
              - https
            path:
              type: string
              pattern: ^/.*$
            retries:
              type: integer
              minimum: 0
            connect_timeout:
              type: integer
              minimum: 0
            read_timeout:
              type: integer
              minimum: 0
            write_timeout:
              type: integer
              minimum: 0
        upstream:
          type: object
          properties:
            hash_on:
              type: string
            hash_on_cookie:
              type: string
            hash_on_cookie_path:
              type: string
            hash_on_header:
              type: string
            hash_fallback_header:
              type: string
            hash_fallback:
              type: string
            slots:
              type: integer
              minimum: 10
            healthchecks:
              type: object
              properties:
                active:
                  type: object
                  properties:
                    concurrency:
                      type: integer
                      minimum: 1
                    timeout:
                      type: integer
                      minimum: 0
                    http_path:
                      type: string
                      pattern: ^/.*$
                    healthy: &healthy
                      type: object
                      properties:
                        http_statuses:
                          type: array
                          items:
                            type: integer
                        interval:
                          type: integer
                          minimum: 0
                        successes:
                          type: integer
                          minimum: 0
                    unhealthy: &unhealthy
                      type: object
                      properties:
                        http_failures:
                          type: integer
                          minimum: 0
                        http_statuses:
                          type: array
                          items:
                            type: integer
                        interval:
                          type: integer
                          minimum: 0
                        tcp_failures:
                          type: integer
                          minimum: 0
                        timeout:
                          type: integer
                          minimum: 0
                passive:
                  type: object
                  properties:
                    healthy: *healthy
                    unhealthy: *unhealthy
```
2. I defined the simple ingress like
```apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  name: api-ingress
  namespace: common
  annotations:
    kubernetes.io/ingress.class: "common"
    plugins.konghq.com: hmac-auth
spec:
  rules:
  - http:
      paths:
      - path: /
        backend:
          serviceName: api-service
          servicePort: 80
```

3. num of nginx worker is 4
```
/ # ps aux|grep nginx
    1 kong       0:00 nginx: master process /usr/local/openresty/nginx/sbin/nginx -p /usr/local/kong -c nginx.conf
   39 kong       0:25 nginx: worker process
   40 kong       0:11 nginx: worker process
   41 kong       0:11 nginx: worker process
   42 kong       0:11 nginx: worker process
   60 root       0:00 grep nginx
```

4. cpu/memory usage. I reboot the pod last day. the memory usage from 200Mi ->1Gi today.
![Screen Shot 2019-04-30 at 10 50 15 AM](https://user-images.githubusercontent.com/9565353/56938706-208c9180-6b36-11e9-9237-abf81f2734cf.png)


Thanks for your help.

**Comment by hbagdi** [comment]
@HsinHeng Thanks for your answers.
We could reproduce this on our side and the underlying issue seems to be related to GC issue with an LRU queue used within the caching layer in Kong.

https://github.com/Kong/kong/pull/4560 seems to have fixed the issue and will be available either in 1.1.x or 1.2.0 release.

@HsinHeng Could you also verify the fix by uninstalling lua-resty-mlcache and then install 2.4.0 version of lua-resty-mlcache and then doing a `kong reload` command inside your container?

**Comment by HsinHeng** [comment]
@hbagdi I have upgraded the lua-resty-mlcache to 2.4.0, and restart kong.

```
/ # luarocks remove  lua-resty-mlcache
/ # luarocks install  lua-resty-mlcache 2.4.0-1
/ # kong reload
/ # luarocks show  lua-resty-mlcache
Warning: The directory '/root/.cache/luarocks' or its parent directory is not owned by the current user and the cache has been disabled. Please check the permissions and owner of that directory. If executing /usr/local/bin/luarocks with sudo, you may want sudo's -H flag.

lua-resty-mlcache 2.4.0-1 - Layered caching library for OpenResty
```

However, memory usage is from 150Mi to 700Mi during these days..
![Screen Shot 2019-05-03 at 10 51 21 AM](https://user-images.githubusercontent.com/9565353/57117914-16ff6580-6d92-11e9-884e-d983bcd0f04e.png)

Please help me to reproduce this issue. Thanks.

**Comment by hishamhm** [comment]
@HsinHeng thank you for all the data, let's keep digging into it. With what frequency is the cronjob posting `/config`?

**Comment by HsinHeng** [comment]
@hishamhm hmm.., it seems to trigger at 4, 7, 11 sec for every minutes.

```
127.0.0.1 - - [03/May/2019:16:30:04 +0000] "POST /config HTTP/1.1" 201 3472 "-" "Go-http-client/1.1"
127.0.0.1 - - [03/May/2019:16:30:07 +0000] "POST /config HTTP/1.1" 201 3472 "-" "Go-http-client/1.1"
127.0.0.1 - - [03/May/2019:16:30:11 +0000] "POST /config HTTP/1.1" 201 3472 "-" "Go-http-client/1.1"
127.0.0.1 - - [03/May/2019:16:40:04 +0000] "POST /config HTTP/1.1" 201 3472 "-" "Go-http-client/1.1"
127.0.0.1 - - [03/May/2019:16:40:07 +0000] "POST /config HTTP/1.1" 201 3472 "-" "Go-http-client/1.1"
127.0.0.1 - - [03/May/2019:16:40:11 +0000] "POST /config HTTP/1.1" 201 3472 "-" "Go-http-client/1.1"
127.0.0.1 - - [03/May/2019:16:50:04 +0000] "POST /config HTTP/1.1" 201 3472 "-" "Go-http-client/1.1"
127.0.0.1 - - [03/May/2019:16:50:07 +0000] "POST /config HTTP/1.1" 201 3471 "-" "Go-http-client/1.1"
127.0.0.1 - - [03/May/2019:16:50:11 +0000] "POST /config HTTP/1.1" 201 3472 "-" "Go-http-client/1.1"
```

**Comment by petreboy14** [comment]
Hi @hishamhm not sure if this is related but we are also running in DB-less mode and I'm noticing a lot of errors like the following in the logs:
```
2019/05/08 15:10:42 [notice] 39#0: *983 [lua] cache.lua:186: purge(): [DB cache] purging (local) cache, client: 127.0.0.1, server: kong_admin, request: "POST /config HTTP/1.1", host: "localhost:8444"
2019/05/08 15:10:42 [error] 39#0: *983 [lua] events.lua:254: post(): worker-events: failed posting event "upstreams" by "crud"; no memory, client: 127.0.0.1, server: kong_admin, request: "POST /config HTTP/1.1", host: "localhost:8444"
2019/05/08 15:10:42 [error] 39#0: *983 [kong] init.lua:249 failed posting crud event for upstreams nil, client: 127.0.0.1, server: kong_admin, request: "POST /config HTTP/1.1", host: "localhost:8444"
2019/05/08 15:10:42 [error] 39#0: *983 [lua] events.lua:345: post(): worker-events: dropping event; waiting for event data timed out, id: 22400, client: 127.0.0.1, server: kong_admin, request: "POST /config HTTP/1.1", host: "localhost:8444"
2019/05/08 15:10:43 [error] 39#0: *983 [lua] events.lua:254: post(): worker-events: failed posting event "upstreams" by "balancer"; no memory, client: 127.0.0.1, server: kong_admin, request: "POST /config HTTP/1.1", host: "localhost:8444"
2019/05/08 15:10:43 [error] 39#0: *983 [lua] handler.lua:512: failed broadcasting upstream create to workers: failed posting event "upstreams" by "balancer"; no memory, client: 127.0.0.1, server: kong_admin, request: "POST /config HTTP/1.1", host: "localhost:8444"
```
Interestingly, when we run one replica of kong it will hover around 2.5GB of memory but when we scale to 3 replicas they all creep up and overnight the leader will get to about 9.8GB (we set a memory limit of 10GB). At this point we get a lot of responses of "failure to get a peer from the ring-balancer" to proxy requests. 
If you don't think this is related I'm happy to open another issue.

**Comment by gpguerra** [comment]
I'm having this same problem with dbless:
2019/05/13 11:24:25 [notice] 37#0: *54655 [lua] cache.lua:186: purge(): [DB cache] purging (local) cache, client: 127.0.0.1, server: kong_admin, request: "POST /config HTTP/1.1", host: "localhost:8444"
2019/05/13 11:24:25 [error] 37#0: *54655 [lua] events.lua:254: post(): worker-events: failed posting event "upstreams" by "crud"; no memory, client: 127.0.0.1, server: kong_admin, request: "POST /config HTTP/1.1", host: "localhost:8444"
2019/05/13 11:24:25 [error] 37#0: *54655 [kong] init.lua:249 failed posting crud event for upstreams nil, client: 127.0.0.1, server: kong_admin, request: "POST /config HTTP/1.1", host: "localhost:8444"
2019/05/13 11:24:25 [error] 37#0: *54655 [lua] events.lua:254: post(): worker-events: failed posting event "upstreams" by "crud"; no memory, client: 127.0.0.1, server: kong_admin, request: "POST /config HTTP/1.1", host: "localhost:8444"
2019/05/13 11:24:25 [error] 37#0: *54655 [kong] init.lua:249 failed posting crud event for upstreams nil, client: 127.0.0.1, server: kong_admin, request: "POST /config HTTP/1.1", host: "localhost:8444"
2019/05/13 11:24:26 [error] 37#0: *54655 [lua] events.lua:345: post(): worker-events: dropping event; waiting for event data timed out, id: 79770, client: 127.0.0.1, server: kong_admin, request: "POST /config HTTP/1.1", host: "localhost:8444"
2019/05/13 11:24:26 [error] 37#0: *54655 [lua] events.lua:345: post(): worker-events: dropping event; waiting for event data timed out, id: 79771, client: 127.0.0.1, server: kong_admin, request: "POST /config HTTP/1.1", host: "localhost:8444"
2019/05/13 11:24:26 [error] 37#0: *54655 [lua] balancer.lua:584: on_target_event(): target create: balancer not found for global-cpp-cvc.services.svc, client: 127.0.0.1, server: kong_admin, request: "POST /config HTTP/1.1", host: "localhost:8444"
2019/05/13 11:24:26 [error] 37#0: *54655 [lua] balancer.lua:584: on_target_event(): target create: balancer not found for broker-partner-azul.services.svc, client: 127.0.0.1, server: kong_admin, request: "POST /config HTTP/1.1", host: "localhost:8444"

And I'm having problem to define our services. I am trying to use kong ingress with CRD, but our plugins and kongIngress get deleted after a time and our services get exposed for a long period until kong re-creates it. We didn't kong admin. just yamls files.

Im using kubernetes 1.12.6

**Comment by MilanDasek** [comment]
@hbagdi - any news? I have exactly same problem. It seems Kong is dropping connections...


thanks

**Comment by hbagdi** [comment]
The connection dropping problem should be solved with atomic reloads in Kong 1.2.

The memory leak issue still isn't solved with Kong 1.2.0rc2. I ran a test over the weekend and it still seems to be happening. We are investigating the issue on our end.

**Comment by hbagdi** [comment]
To share an update: https://github.com/Kong/kong/pull/4733

**Comment by HsinHeng** [comment]
It's good news to found the root cause.
We are appreciated to look forward for your upgrade.

**Comment by ismailbaskin** [comment]
It has been solved by [Kong 1.2.1](https://github.com/Kong/kong/blob/master/CHANGELOG.md#121)
